### ISSUE TITLE
Users: Please check in here to let me know you are here and give me an initial report about this driver...

### ISSUE THREAD

**Issue by morrownr** [morrownr/rtl8852bu-20240418#2]
I would appreciate it if you would post a message here so that I know that you are using the driver. Also, please give an initial report of the driver. If you find a bug that needs to be fixed, please post a new issue.

Thanks and enjoy.

**Comment by mrc99** [comment]
Hi @morrownr. Just checking in. Have cloned the driver and compiled OK under Gentoo with kernel 6.12.7. UDP iperf3 results below. Note that this is on a USB2 port as the USB3.0 ports on this old mb cause USB device lockups under load. I've ordered a PCIe USB3.2 to have a play with so will re-post results when received.

One minor difference noted from previous driver is that make install always edits /etc/modprobe.d/8852bu.conf whereas the old driver make install didn't. Is this intended behaviour?

This adapter is in use daily with no errors/issues spotted to date but shout if you want any specific tests carried out.

iperf3 -c 192.168.101.203 -u -bidir --bind-dev wlan0 -V

iperf 3.17.1
Linux lucy 6.12.7 #1 SMP PREEMPT Fri Dec 27 18:28:07 GMT 2024 x86_64
Control connection MSS 1448
Setting UDP block size to 1448
Time: Fri, 27 Dec 2024 22:00:44 GMT
Connecting to host 192.168.101.203, port 5201
Cookie: wj2qbee2d3lw4ne5ygxbv5alvuiluczbf2g3
[  5] local 192.168.101.78 port 43151 connected to 192.168.101.203 port 5201
Starting Test: protocol: UDP, 1 streams, 1448 byte blocks, omitting 0 seconds, 10 second test, tos 0
[ ID] Interval           Transfer     Bitrate         Total Datagrams
[  5]   0.00-1.00   sec  26.1 MBytes   219 Mbits/sec  18911
[  5]   1.00-2.00   sec  25.7 MBytes   216 Mbits/sec  18640
[  5]   2.00-3.00   sec  27.2 MBytes   228 Mbits/sec  19692
[  5]   3.00-4.00   sec  27.2 MBytes   229 Mbits/sec  19728
[  5]   4.00-5.00   sec  27.3 MBytes   229 Mbits/sec  19800
[  5]   5.00-6.00   sec  27.4 MBytes   230 Mbits/sec  19812
[  5]   6.00-7.00   sec  27.9 MBytes   234 Mbits/sec  20192
[  5]   7.00-8.00   sec  27.0 MBytes   226 Mbits/sec  19560
[  5]   8.00-9.00   sec  26.6 MBytes   223 Mbits/sec  19272
[  5]   9.00-10.00  sec  27.1 MBytes   228 Mbits/sec  19656
- - - - - - - - - - - - - - - - - - - - - - - - -
Test Complete. Summary Results:
[ ID] Interval           Transfer     Bitrate         Jitter    Lost/Total Datagrams
[  5]   0.00-10.00  sec   270 MBytes   226 Mbits/sec  0.000 ms  0/195263 (0%)  sender
[  5]   0.00-10.02  sec   261 MBytes   218 Mbits/sec  0.050 ms  6282/195247 (3.2%)  receiver
CPU Utilization: local/sender 16.2% (1.3%u/14.9%s), remote/receiver 25.0% (2.9%u/22.1%s)

iperf Done.

**Comment by morrownr** [comment]
Hi @mrc99 

> One minor difference noted from previous driver is that make install always edits /etc/modprobe.d/8852bu.conf whereas the old driver make install didn't. Is this intended behaviour?

It is the intended behavior. If you see a problem with it, let's discuss it and see what we can do.

> This adapter is in use daily with no errors/issues spotted to date but shout if you want any specific tests carried out.

I have tested managed mode extensively and I am testing AP mode right now. I am finding no bugs in either mode but I do need testers to work the P2P modes and IBSS if you want to take a look at one or both.

My experience so far is that this driver is MUCH better than the previous driver.

**Comment by s1shed** [comment]
Hi, checking in. I've been testing in IBSS mode and not having success thus far. 

On systems running Kali Linux (Linux 6.11.2), Debian Sid running 6.12.6, and Debian stable running 6.1.0, the command `iw dev DEVICENAME link` returns:

```
Not connected.
```
after trying to join a network using IBSS.


Steps performed:

```
ip link set dev wlan1 down
iw wlan1 set type ibss
ip link set dev wlan1 up
iw wlan1 ibss join test 2412
iw dev wlan1 link
```

This procedure works with with my Alfa AWUS036NH.

I'll play with this more tomorrow to see if I can get it to work.

(Managed mode works very reliably.)

**Comment by morrownr** [comment]
> I'll play with this more tomorrow to see if I can get it to work.

Thanks. No hurry. These Realtek drivers do not follow normal ways of doing things in some modes so you may need to do some searching.

> (Managed mode works very reliably.)

Yes and AP mode does as well. That is a big improvement over the old driver.

**Comment by s1shed** [comment]
Follow-up #2:

I've tried this on the same systems from last night's follow-up, setting up IBSS using `wpa_supplicant` instead of with `iw. Since the behaviour seems to be the same I'll report based on one of the systems, the one running Kali Linux with all current updates applied.

```
: ---------------------------
: Please copy and post all below lines when reporting an issue!
: ---------------------------
: install-driver.sh v20241208
: x86_64 (kernel architecture)
: x86_64 (architecture to send to gcc)
: 4/4 (in-use/total processing units)
: 7892444 (total system memory)
: 6.11.2-amd64 (kernel version)
: gcc-14 (version of gcc used to compile the kernel)
: gcc (Debian 14.2.0-8) 14.2.0
: dkms-3.0.13
: SecureBoot disabled
: Linux Wireless Regulatory Settings:
global
country FI: DFS-ETSI
phy#0 (self-managed)
country FI: DFS-UNSET
: Info: https://docs.kernel.org/networking/regulatory.html
: Info: https://www.marcusfolkesson.se/blog/linux-wireless-regulatory/
: ---------------------------
```

Contents of my `wpa_supplicant` config:

```
#  wpa_supplicant -B -i wlan1 -c adhoc.conf

ctrl_interface=/run/wpa_supplicant GROUP=root

network={
        ssid="testnet"
        mode=1
        frequency=2432
        proto=RSN
        key_mgmt=WPA-PSK
        pairwise=CCMP
        group=CCMP
        psk="testnet1234"
}
```

Starting `wpa_supplicant` using the command-line and config provided above results in a kernel crash.

```
Dec 30 14:17:13 kali kernel: ------------[ cut here ]------------
Dec 30 14:17:13 kali kernel: WARNING: CPU: 3 PID: 1640 at net/wireless/ibss.c:37 __cfg80211_ibss_joined+0x19b/0x1b0 [cfg80211]
Dec 30 14:17:13 kali kernel: Modules linked in: ccm snd_seq_dummy snd_hrtimer snd_seq snd_seq_device overlay qrtr 8852bu(OE) btusb uvcvideo btrtl videobuf2_vmalloc uvc btintel btbcm videobuf2_memops videobuf2_v4l2 btmtk sunrpc videodev bluetooth cdc_mbim cdc_wdm videobuf2_common binfmt_misc cdc_acm mc snd_sof_pci_intel_skl snd_sof_intel_hda_generic soundwire_intel soundwire_generic_allocation soundwire_cadence snd_sof_intel_hda_common snd_sof_intel_hda_mlink snd_sof_intel_hda snd_sof_pci snd_sof_xtensa_dsp intel_rapl_msr snd_sof intel_rapl_common intel_uncore_frequency intel_uncore_frequency_common intel_pmc_core_pltdrv intel_pmc_core intel_vsec nls_ascii snd_sof_utils pmt_telemetry soundwire_bus nls_cp437 pmt_class vfat fat x86_pkg_temp_thermal intel_powerclamp snd_soc_avs coretemp snd_soc_hda_codec iwlmvm kvm_intel snd_soc_skl ext4 mac80211 snd_soc_hdac_hda snd_hda_ext_core snd_soc_sst_ipc snd_soc_sst_dsp crc16 mbcache jbd2 snd_soc_acpi_intel_match snd_hda_codec_hdmi libarc4 snd_soc_acpi kvm snd_soc_core iwlwifi tps6598x
Dec 30 14:17:13 kali kernel:  snd_hda_codec_realtek snd_hda_codec_generic typec snd_hda_scodec_component snd_compress mei_pxp mei_hdcp snd_pcm_dmaengine snd_hda_intel snd_intel_dspcfg snd_intel_sdw_acpi snd_hda_codec iTCO_wdt snd_ctl_led rapl snd_hda_core intel_cstate cfg80211 intel_pmc_bxt intel_uncore snd_hwdep snd_pcm iTCO_vendor_support think_lmi thinkpad_acpi firmware_attributes_class pcspkr watchdog wmi_bmof intel_wmi_thunderbolt ee1004 snd_timer mei_me nvram mei sparse_keymap platform_profile intel_xhci_usb_role_switch intel_pch_thermal roles serial_multi_instantiate snd soundcore rfkill ac acpi_pad joydev serio_raw evdev vmwgfx drm_ttm_helper msr efi_pstore configfs nfnetlink efivarfs ip_tables x_tables autofs4 btrfs blake2b_generic xor raid6_pq libcrc32c crc32c_generic dm_crypt dm_mod hid_logitech_hidpp hid_logitech_dj hid_generic usbhid hid cdc_ncm cdc_ether usbnet mii i915 crct10dif_pclmul crc32_pclmul drm_buddy crc32c_intel i2c_algo_bit drm_display_helper cec ghash_clmulni_intel rc_core xhci_pci ttm sha512_ssse3 xhci_hcd
Dec 30 14:17:13 kali kernel:  drm_kms_helper rtsx_pci_sdmmc sha256_ssse3 mmc_core usbcore i2c_i801 nvme intel_lpss_pci sha1_ssse3 psmouse e1000e drm nvme_core i2c_smbus rtsx_pci intel_lpss idma64 usb_common battery video wmi button aesni_intel gf128mul crypto_simd cryptd
Dec 30 14:17:13 kali kernel: CPU: 3 UID: 0 PID: 1640 Comm: kworker/u16:7 Tainted: G           OE      6.11.2-amd64 #1  Kali 6.11.2-1kali1
Dec 30 14:17:13 kali kernel: Tainted: [O]=OOT_MODULE, [E]=UNSIGNED_MODULE
Dec 30 14:17:13 kali kernel: Hardware name: LENOVO 20HN0012MX/20HN0012MX, BIOS R0IET72W (1.50 ) 11/08/2023
Dec 30 14:17:13 kali kernel: Workqueue: cfg80211 cfg80211_event_work [cfg80211]
Dec 30 14:17:13 kali kernel: RIP: 0010:__cfg80211_ibss_joined+0x19b/0x1b0 [cfg80211]
Dec 30 14:17:13 kali kernel: Code: c7 45 b8 00 00 00 00 66 89 45 b6 e8 5f 3f c9 ca e9 d8 fe ff ff 0f 0b e9 d1 fe ff ff 0f 0b e9 44 ff ff ff 0f 0b e9 5c ff ff ff <0f> 0b e9 bc fe ff ff 0f 0b e8 67 ad d5 ca 0f 1f 80 00 00 00 00 90
Dec 30 14:17:13 kali kernel: RSP: 0018:ffffb031482cfda0 EFLAGS: 00010246
Dec 30 14:17:13 kali kernel: RAX: 0000000000000002 RBX: ffff9e40419eb000 RCX: 0000000000000001
Dec 30 14:17:13 kali kernel: RDX: 0000000000000001 RSI: 00000000fffffe00 RDI: ffffffffc132d5da
Dec 30 14:17:13 kali kernel: RBP: ffffb031482cfdf0 R08: 0000000000000000 R09: 0000000000000001
Dec 30 14:17:13 kali kernel: R10: ffff9e40663a50c0 R11: 0000000000000001 R12: ffff9e405281b000
Dec 30 14:17:13 kali kernel: R13: ffff9e406264e818 R14: 0000000000000000 R15: dead000000000100
Dec 30 14:17:13 kali kernel: FS:  0000000000000000(0000) GS:ffff9e416a780000(0000) knlGS:0000000000000000
Dec 30 14:17:13 kali kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
Dec 30 14:17:13 kali kernel: CR2: 00007f38834d0490 CR3: 0000000007c22006 CR4: 00000000003706f0
Dec 30 14:17:13 kali kernel: Call Trace:
Dec 30 14:17:13 kali kernel:  <TASK>
Dec 30 14:17:13 kali kernel:  ? __cfg80211_ibss_joined+0x19b/0x1b0 [cfg80211]
Dec 30 14:17:13 kali kernel:  ? __warn.cold+0x8e/0xe8
Dec 30 14:17:13 kali kernel:  ? __cfg80211_ibss_joined+0x19b/0x1b0 [cfg80211]
Dec 30 14:17:13 kali kernel:  ? report_bug+0xff/0x140
Dec 30 14:17:13 kali kernel:  ? handle_bug+0x3c/0x80
Dec 30 14:17:13 kali kernel:  ? exc_invalid_op+0x17/0x70
Dec 30 14:17:13 kali kernel:  ? asm_exc_invalid_op+0x1a/0x20
Dec 30 14:17:13 kali kernel:  ? __cfg80211_get_bss+0x26a/0x2d0 [cfg80211]
Dec 30 14:17:13 kali kernel:  ? __cfg80211_ibss_joined+0x19b/0x1b0 [cfg80211]
Dec 30 14:17:13 kali kernel:  ? cfg80211_process_wdev_events+0x122/0x1a0 [cfg80211]
Dec 30 14:17:13 kali kernel:  cfg80211_process_wdev_events+0x122/0x1a0 [cfg80211]
Dec 30 14:17:13 kali kernel:  cfg80211_process_rdev_events+0x2a/0x40 [cfg80211]
Dec 30 14:17:13 kali kernel:  cfg80211_event_work+0x29/0x40 [cfg80211]
Dec 30 14:17:13 kali kernel:  process_one_work+0x174/0x330
Dec 30 14:17:13 kali kernel:  worker_thread+0x252/0x390
Dec 30 14:17:13 kali kernel:  ? __pfx_worker_thread+0x10/0x10
Dec 30 14:17:13 kali kernel:  kthread+0xcf/0x100
Dec 30 14:17:13 kali kernel:  ? __pfx_kthread+0x10/0x10
Dec 30 14:17:13 kali kernel:  ret_from_fork+0x31/0x50
Dec 30 14:17:13 kali kernel:  ? __pfx_kthread+0x10/0x10
Dec 30 14:17:13 kali kernel:  ret_from_fork_asm+0x1a/0x30
Dec 30 14:17:13 kali kernel:  </TASK>
Dec 30 14:17:13 kali kernel: ---[ end trace 0000000000000000 ]---
``` 

For what it's worth, this has been seen in USB2 and USB3 modes, verified with the output from `lsusb`.

Note: The [old driver](https://github.com/morrownr/rtl8852bu) doesn't work for me in IBSS mode either, but there's no kernel crash. 

Please let me know if there's any other information that I could provide that would be of use.

**Comment by morrownr** [comment]
I appreciate you trying. Since this is a rarely used capability anyway, I am going to remove it as a capability from the README and only come back to the issue if I need to do so.

**Comment by GrandMoff100** [comment]
Hey @morrownr I'm using your driver on my OpenMediaVault installation with an Asus USB-AX55 Nano.

Works like a charm. Thank you so much!

**Comment by morrownr** [comment]
Hi @GrandMoff100 

Thanks for the report. 

> (P.S. you should pin this issue so more people see it)

It is now pinned. I added this issue as I have not had much luck finding testers.

**Comment by pnrhub** [comment]
Just installed this driver on a Star64 src board (a RISCV board made by Pine64, very similar to the VisionFive2 board). The OS is Debian Trixie (i.e. kernel 6.12). The build and install went smoothly, no issues to report. Instructions were clear. The wifi module appears to work just fine. Many thanks for providing this.

One problem: on each boot the wifi device gets a new MAC address. Is that intended behaviour? How can I configure (or patch) to set a fixed MAC address -- this is more convenient with network manager and also for 'sticky' dhcp on the router.

**Comment by morrownr** [comment]
@pnrhub 

Thanks for the report. It is especially important for me to see the results of testing on RISCV based systems as I do not have one to test with. I have setup AUTODETECT capability so that the installation can autodetect the CPU and compile for the correct platform. Evidently it works for RISCV. Good stuff.

> How can I configure (or patch) to set a fixed MAC address?

I do not think the driver is doing this. I did a quick search and here is what I came up with:
 
To stop a Linux wireless device from getting a new MAC address on each reboot, you need to disable MAC address randomization within your network settings, typically by accessing the NetworkManager configuration and setting the "wifi.cloned-mac-address" option to "stable" or specifying a fixed MAC address for your wireless interface; this can usually be done through your system's graphical network settings or by editing the /etc/NetworkManager/NetworkManager.conf file.

**Comment by pnrhub** [comment]
Thanks. Diving into nmcli at the moment; not resolved yet but I'll get there.

Also noticed the following in the boot log:
[   28.071392] Bluetooth: hci0: RTL: firmware file rtl_bt/rtl8852bu_fw not found
(was probably always there, but I did not notice before).

Did not test bluetooth, but wifi appears to work fine.

**Comment by morrownr** [comment]
> [ 28.071392] Bluetooth: hci0: RTL: firmware file rtl_bt/rtl8852bu_fw not found

Does your adapter have BT support?

**Comment by sateuwdie** [comment]
Works bad on some pc. I have tried in a hostapd AP and give me disconnection under heavy load

https://github.com/morrownr/rtl8852bu-20240418/issues/8

In another pc (asus A88Xm-e) used as client don't connect, revert to old driver connection is perfect.

**Comment by morrownr** [comment]
@sateuwdie 

I responded to #8 

> In another pc (asus A88Xm-e) used as client don't connect, revert to old driver connection is perfect.

Interesting. How did you set up the `options` ?

**Comment by pnrhub** [comment]
> > [ 28.071392] Bluetooth: hci0: RTL: firmware file rtl_bt/rtl8852bu_fw not found
> 
> Does your adapter have BT support?

Yes.

It would seem that the `/lib/firmware/rtl_bt/rtl8852bu_fw.bin` file is not present following the install.

Still working on the stable MAC issue, I've not been able to get this right yet.

**Comment by morrownr** [comment]
It could be that there is a bad setting somewhere. These Realtek out-of-kernel drivers are not really made for us poor mortals trying to run an adapter. Many default settings have to be changed and these WiFi 6 drivers are incredibly complicated.

**Comment by pnrhub** [comment]
> It could be that there is a bad setting somewhere. These Realtek out-of-kernel drivers are not really made for us poor mortals trying to run an adapter. Many default settings have to be changed and these WiFi 6 drivers are incredibly complicated.

Found it. The diagnosis and solution is described here:
https://forum.radxa.com/t/making-wlan0-mac-address-static-fixed/5445

For NetworkManager to do its job, it needs the hardware to provide a fixed MAC address. In this case the hardware does not provide that and we have to rely on configuring it in the driver, via the `rtw_initmac` setting in the config file. I would recommend that the install script is changed to provide such a fixed "pseudo hardware" MAC.

Currently, during installation */etc/modprobe.d/8852bu.conf* is initialised with:
`options 8852bu rtw_switch_usb_mode=0`
I recommend this is changed to for example:
`options 8852bu rtw_switch_usb_mode=0 rtw_initmac="00:E0:4C:01:02:03"`

Of course it is better to have it randomised during install. One way could be to fix the first three bytes and to assign the last three bytes at random.

With a pseudo hardware MAC in place, NetworkManager works as expected. The added benefit is that with this addition the device name will also become stable/predictable in Linux 6.x

**Comment by pnrhub** [comment]
> It would seem that the /lib/firmware/rtl_bt/rtl8852bu_fw.bin file is not present following the install.

I guess Realtek is not providing a RISCV build of the rtl8852bu_fw.bin binary in its support package. In a few weeks Starfive are coming out with a new package for the FV2 / Star64, which is hopefully based on the same upstream code base as this driver. If so, I should be able to use this.

For Intel and Arm, there should be many sources to find a usable Linux firmware module file, e.g. here for Arm
https://github.com/armbian/firmware/tree/master/rtl_bt

Maybe some notes should be added to the README about this.

**Comment by morrownr** [comment]
> I guess Realtek is not providing a RISCV build of the rtl8852bu_fw.bin binary ...

There is no such thing as a RISCV or x86_64 or AMD64 or ARM64 firmware file. It is a binary file.

In this type of driver, the firmware file(s) is included here in the driver. It may be that there is something different than previous driver of this type and I have not set the correct setting to enable BT support. It is not obvious to me what should be done at this point. You are the first person to bring this issue so we'll have to figure out.

> For NetworkManager to do its job, it needs the hardware to provide a fixed MAC address...

There has to be a better solution than adding a module parameter. Let check the source and the Makefile to see if we can do this in a way that won't confuse people. Linux does not have to be complicated. The in-kernel drivers are not this way. It is like comparing an automatic transmission to a standard transmission.

I appreciate your reports. You have hit on 2 issues that have not come up previously. Let's see if we can sort this out.

**Comment by morrownr** [comment]
> There is no such thing as a RISCV or x86_64 or AMD64 or ARM64 firmware file. It is a binary file.

After thinking about it, that may not be true.

I'm sick today.

**Comment by morrownr** [comment]
I was looking at the default value for rtw_initmac= and it is null. The adapter should be using the mac address in efuse. I think I have the setting now. I need to test.

**Comment by morrownr** [comment]
I have merged a patch for you to test.

Please run the following in the driver directory:

$ git pull
$ sudo sh install-driver.sh

When the script asks about module parameters (options), do not add:

rtw_initmac="00:E0:4C:01:02:03"

See if the patch helped. If I am reading code correctly, the patch should allow the driver to read the mac from ufuse now... which is what the default should have been in the first place.

Let me know.

FYI: Still looking at the bluetooth issue.

**Comment by pnrhub** [comment]
The patch did not change the behaviour: MAC still not stable. Sorry.

I am not sure the module on my board (see https://files.pine64.org/doc/datasheet/star64/Fn-Link_6252B-UUB_datasheet_V1.1_20220329.pdf) has fuses or prom (unless that is on the 8852BU chip itself). Even if there are fuses, maybe they are not set.

The workaround via a driver parameter is good enough for my use case, but I understand it would not cover a use case where two 8852-based Wifi dongles are inserted into the same machine.

====

Separately, for bluetooth I did install the Armbian firmware files and the result was mixed. The boot log shows it being loaded, but there also seems to be an error with some symbol not resolved. Yet, bluetooth appears to initialise and work: in a simple test it could pair with my phone. However, the connection was dropped within a few seconds and can then not be re-established. I've never really used bluetooth on linux, so I am not sure if I'm doing something wrong or whether this is even supposed to work.

My (rather superficial) hypothesis now would be that these files are not architecture dependent, but are matched to a specific version of the overall driver.

**Comment by morrownr** [comment]
@pnrhub 

> The patch did not change the behaviour: MAC still not stable.

Thanks for letting me know. Some else for my to-do list.

> The workaround via a driver parameter is good enough for my use case, but I understand it would not cover a use case where two 8852-based Wifi dongles are inserted into the same machine.

It is a long standing issue that you can't run two Realtek based adapters with the same chipset in the same system. I just have to recommend Mediatek adapters as that it not a problem with them.

> Separately, for bluetooth I did install the Armbian firmware files and the result was mixed.

It appears that these WiFi 6 chips have separate wifi and bt drivers. Can you run the following so that I can see the ID for the bluetooth support?

$ lsusb

Like a lot of things Realtek, it may be hard to find a current copy of this bluetooth driver. The reason I can keep very recent versions of some of these Realtek drivers is because I work with a couple of makers and retailers and they have access to the code for the products they buy. None of my suppliers sell a multi-function adapter. Does the supplier of your module make Linux drivers available or can they get them for you? It would not hurt to ask but make sure you ask for both wifi and bluetooth drivers.

**Comment by morrownr** [comment]
@pnrhub 

I was doing some research on your bluetooth issue. It seems that support for the rtl8852be originally came out as an out-of-kernel driver and its bluetooth support was a separate out-of-kernel driver. Since that time, Realtek's team that handles chips for cards, internal stuff, has started supporting in-kernel drivers for both... certainly for wifi and I think for bluetooth as well. It also appears that the bluetooth support for the rtl8852be and rtl8852bu are the same so my guess is that bluetooth should be working on the most recent releases. It is possible that distros mess up the firmware. Here is the location of the bt firmware files:

 https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/rtl_bt

It would seem that rtl8852bu_fw.bin and rtl8852btu_fw.bin would need to be placed in /lib/firmware/rtl_bt

You might try using these files as they are from the official repo and may be more recent that what you have. Beyond that, things are still fuzzy but I think I am telling you how it is this bt issue should fix itself over time. What distro and kernel are you using? 

You are the first person I have seen with a multi-function adapter/module with a rtl8852bu chipset. That link to the details did confirm that your wifi is limited to USB2, which should be expected with bt onboard. If you are not getting the speed results you are expecting, don't blame it on the driver.

You probably don't want to hear this but on thing I strongly recommend to Linux users is to above usb adapters/modules that come with bluetooth support. It make for interference problems.

**Comment by pnrhub** [comment]
>It appears that these WiFi 6 chips have separate wifi and bt drivers. Can you run the following so that I can see the ID for the bluetooth support?
```
# lsusb
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 001 Device 002: ID 1a40:0101 Terminus Technology Inc. Hub
Bus 001 Device 003: ID 1a40:0101 Terminus Technology Inc. Hub
Bus 001 Device 004: ID 0bda:b852 Realtek Semiconductor Corp. 802.11ax WLAN Adapter
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
```
It would seem that the WiFi side and the Bluetooth side share a single device ID and that they work via separate interfaces (as defined in the USB protocol spec).

**Comment by pnrhub** [comment]
> You probably don't want to hear this but on thing I strongly recommend to Linux users is to above usb adapters/modules that come with bluetooth support. It make for interference problems.

Thanks for all the info. My purpose for the Star64 board is mostly to experiment (when I have time) with a RiscV-based board, I'm not at all worried if something doesn't work. With some further experimentation, it would seem that the Armbian file does bring the Bluetooth side up properly, and the connection problems I mentioned are higher up the stack and not in the driver. I'm not normally working with Bluetooth, so I was investigating this for you, to give you the user feedback you asked for.

Currently investigating Debian Trixie on RiscV. Last year I was using this Yocto stack, that includes a rtl8852bu driver, I think based on similar upstream code:
https://github.com/Fishwaldo/Star64_linux/tree/Star64_devel/drivers/staging/rtl8852bu

Starfive have announced they'll have a new BSP release by the end of this month, I'll share a  link once it arrives. It should have a driver for this same module as well.

In summary: I'm perfectly fine with where things stand and thank you for providing this repo and for your kind help in the past days.
